README.rst
============================
Curses Based Clock & Bubbles
============================

This repository contains codes of ``clock.c`` and ``bubbles.c``, which was
written by Martin Sullivan in 80s and 90s, you can download the original
tarball, in K&R C, from `A Curses Based Clock`_.

.. _A Curses Based Clock: http://www.zois.co.uk/tn/tn-1991-01-01.html

.. contents:: **Contents**
   :local:


Installation
============

You need a C compiler that supports C99 and POSIX.1-2008.

By default, both ``bubbles`` and ``clock`` binaries will be installed under
``/usr/local``, you can use ``prefix=DIR`` and/or specify a target to only
install either binary.

.. code:: sh

  $ make [prefix=$HOME/.local] [bubbles|clock] install[-strip]

Use ``install-strip`` if you want stripped binaries. For uninstallation, use
``uninstall`` target.


``clock.c``: *ASCII analog clock in terminal*
=============================================

.. figure:: https://bytebucket.org/livibetter/clock/raw/tip/clock.gif

   Some options

.. figure:: https://bytebucket.org/livibetter/clock/raw/tip/clock.roman.png

   Roman clock face

Options
-------

The usage is::

  clock [OPTION] [TITLE]

+------------------------+----------------------------------------+
| option and argument    | description                            |
+========+===============+========================================+
| ``-s`` | ``--second``  | display second hand                    |
|        | = ``STYLE``   | , three different Styles_              |
+--------+---------------+----------------------------------------+
| ``-r`` | ``--roman``   | use Roman numeral on square clock face |
+--------+---------------+----------------------------------------+
| ``-f`` |               | display date and day of week           |
+--------+---------------+----------------------------------------+
| ``-d`` | ``--date``    | display full date/time                 |
+--------+---------------+----------------------------------------+
| ``-h`` | ``--help``    | display help message and exit          |
+--------+---------------+----------------------------------------+
| ``-v`` | ``--version`` | display version and exit               |
+--------+---------------+----------------------------------------+
| ``TITLE``              | use custom text instead of "(\|./)"    |
+------------------------+----------------------------------------+

Styles
------

0: ``OFF`` (default)
  Do not display second hand

1: ``BLOB`` (default on)
  Display as a blob

2: ``LINE``
  Display the normal clock second hand

Controls
--------

+-------+----------------------------------+
| key   | description                      |
+=======+==================================+
| ``q`` | quit the clock                   |
+-------+----------------------------------+
| ``s`` | toggle second hand               |
+-------+----------------------------------+
| ``r`` | toggle Roman numeral square face |
+-------+----------------------------------+
| ``f`` | toggle date and day of week      |
+-------+----------------------------------+
| ``d`` | toggle full date and time        |
+-------+----------------------------------+


``bubbles.c``: *Animated ASCII bubbles in terminal*
===================================================

.. figure:: https://bytebucket.org/livibetter/clock/raw/tip/bubbles.gif

Options
-------

The usage is::

  bubbles [OPTION]

+------------------------+-------------------------------+---------+
| option                 | description                   | default |
+========+===============+===============================+=========+
| ``-d`` | ``--delay``   | update delay in milliseconds  | 500     |
+--------+---------------+-------------------------------+---------+
| ``-h`` | ``--help``    | display help message and exit |         |
+--------+---------------+-------------------------------+---------+
| ``-v`` | ``--version`` | display version and exit      |         |
+--------+---------------+-------------------------------+---------+

Controls
--------

+-------+------------------+
| key   | description      |
+-------+------------------+
| ``q`` | quit the bubbles |
+-------+------------------+


History
=======

In March, 2014 [#ts]_, I (@livibetter, Yu-Jie Lin) accidentally found the
*clock*, but couldn't compile using GCC 4.7.3. The code, ``clock.c``, was
written in K&R C, although GCC seems to have options_ to compile K&R C
(non-ISO), ``-E -traditional``, Gnulib ain't happy with that, asking for an ISO
C compiler to use its headers.

.. _options: https://gcc.gnu.org/onlinedocs/gcc-3.1/gcc/Incompatibilities.html

Updating the code probably is the only way to resolve those errors and
warnings, but I want to keep the most parts untouched and are able to be
compiled in modern environment. With my limited knowledge of Standard and K&R
C, I still managed to achieve my goal. Now, it could be ``make`` without
spitting out any errors and warnings. Note that -- besides ``clock.c`` -- there
is also a ``bubbles.c``, which is fixed as well.

Originally, this was only meant to fix errors and warnings, but in mid-July,
decisions were made to transitioning into a C99 standard and
POSIX.1-2008-conformant project.

After I created this repository, I emailed Martin Sullivan for some questions
and soon got an reply with a bit of history about this clock:

  Bubbles is by far the oldest of them. I wrote bubbles sometime between
  1984 and 1986 when I was teaching myself curses(3). Curses was
  important back then for it gave a fast, independent, character-based
  interface to all the VT100-like terminals we used to use. Bubbles was
  inspired by rain, a curses-based toy that was distributed with a
  variety of standard UNIXs and produced an effect not unlike rain drops
  on your screen. Subsequently bubbles has been cleaned up, and had a
  couple of bugs removed. It's the version you see now, but it's full of
  bad-practice 'C' too.

  The clock program is whole lot older, but inspired by earlier work. At
  the time I was working for ICL, in the early '90s, on a then new
  machine, a Sparc/VME-bus/SCSI based mini-computer that would become
  the DRS/6000. On display, somebody had written a rather crude script
  involving clear(1), echo(1) and date(1) to repetitively print the time
  on the middle of the console screen on the demo machine. I thought I
  could do better with a lunch-time hack. The clock program went through
  a number of iterations after that, but the code was ugly and contained
  some disgusting hacks. Since it went through a SVR4 C compiler
  relatively cleanly nothing much was done with it. I later wrote it up
  as a `Technical Note`__, giving it a whimsical date from around the time I
  wrote it. It will still compile for me with a 4.4.3 compiler under
  Linux, although not without several pages of warnings about my lax
  approached to C-language standards.

__ `A Curses Based Clock`_

`Version 0.1.0`_ was modified from the original code for modern compiling
environment.

.. _Version 0.1.0: https://bitbucket.org/livibetter/clock/commits/tag/v0.1.0


Copyright
=========

Licensed under the GNU General Public License Version 1, see |COPYING|_::

  Copyright (C) 2014 Alexandre Dantas
  Copyright (C) 2014 Yu-Jie Lin
  Copyright (C) 1993 ZOIS Ltd.
  Copyright (C) 1984-1986 Martin Sullivan

.. |COPYING| replace:: ``COPYING``
.. _COPYING: https://bitbucket.org/livibetter/clock/src/tip/COPYING

----

.. [#ts] The timestamp is ``2014-03-19T07:53:28Z``


NEWS
===========================
NEWS for 0.3.1 (2014-09-29)
===========================

Changes since 0.3.0 (2014-09-19)
================================

[bubbles]

* fix bubbles randomness

Changes since 0.2.0 (2014-08-09)
================================

[clock]

* second hand can be set to OFF (default), BLOB (default on), or LINE, e.g.
  -s0, -sBLOB, or --second=LINE.


clock.h
/*
 * Copyright (C) 2014 Yu-Jie Lin
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#include <stdbool.h>

#include "common.h"


#define PROGRAM_NAME "clock"
#define HELP_MESSAGE "\
Usage: " PROGRAM_NAME " [OPTION]...\n\
ASCII analog clock in terminal\n\
\n\
  -s, --second[=STYLE]      display second hand, STYLE can be either number\n\
                            or name (Default: 0: OFF)\n\
  -r, --roman               display square face with Roman numerals\n\
  -d, --date                display digital date and time\n\
  -f                        display date and day of week\n\
  -h, --help                display this help and exit\n\
  -v, --version             display version and exit\n\
\n\
Also, press any of the keys above to change the behavior\n\
while the clock is running.\n\
For example, press 's' to cylce through the second hand styles.\n\
\n\
STYLE:\n\
  0: OFF\n\
  1: BLOB\n\
  2: LINE\n\
\n\
Report bugs to <" URL_ISSUES ">\n\
Home page: <" URL_HOMEPAGE ">\n\
For complete documentation, run: man " PROGRAM_NAME

#define DEFAULT_TITLE "(|./)"

#define HALF_Y ((double) (LINES - 1) / (double) 2.0)
#define HALF_X ((double) (COLS  - 1) / (double) 2.0)
// normal presentation of screen, ratio y : x on normal screen
#define ASPECT ((double) (7.5 / 10.0))

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif


struct cartesian {
    double x;
    double y;
};

struct icartesian {
    int ix;
    int iy;
};

enum second_style {
    SS_OFF = 0,
    SS_BLOB,
    SS_LINE,
    SS_BEGIN      = SS_OFF,
    SS_END        = SS_LINE,
    SS_DEFAULT    = SS_OFF,
    SS_DEFAULT_ON = SS_BLOB
};

struct clock_mode {
    enum second_style second;
    bool roman;
    bool day_date;
    bool digital;
};


void plot (struct cartesian *start, struct cartesian *end, char c);
void putline (int x0, int y0, int x1, int y1, char c);


common.c
/*
 * Copyright (C) 2014 Yu-Jie Lin
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#include <curses.h>
#include <stdio.h>
#include <stdlib.h>

#include "common.h"


void
print_version ()
{
    printf("%s %s\n", program_name, VERSION);
    puts("\
Copyright (C) 2014 Authors of Clock project\n\
License GPLv1+: GNU GPL version 1 or later <https://www.gnu.org/licenses/gpl-1.0.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.");
}


void
usage (int status)
{
    if (status != EXIT_SUCCESS)
        fprintf(stderr, TRY, program_name);
    else
        puts(help_message);

    exit(status);
}


void
abort_handle (int signum GCC_UNUSED)
{
    move(LINES - 1, 0);
    refresh();
    endwin();
    exit(EXIT_SUCCESS);
}


common.h
/*
 * Copyright (C) 2014 Yu-Jie Lin
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#ifndef COMMON_H
#define COMMON_H

#define VERSION "0.3.1"

#define URL_HOMEPAGE  "https://bitbucket.org/livibetter/clock"
#define URL_ISSUES    "https://bitbucket.org/livibetter/clock/issues"

#define TRY "Try '%s --help' for more information.\n"

#ifdef __GNUC__
// curses.h or other sources may already define
#undef  GCC_UNUSED
#define GCC_UNUSED __attribute__((unused))
#else
#define GCC_UNUSED /* nothing */
#endif


extern char *program_name;
extern char *help_message;


void print_version ();
void usage (int status);

void abort_handle (int signum GCC_UNUSED);

#endif


bubbles.c
/*
 * Animated ASCII bubbles in terminal
 * Copyright (C) 2014 Alexandre Dantas
 * Copyright (C) 2014 Yu-Jie Lin
 * Copyright (C) 1984-1986 Martin Sullivan
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <curses.h>
#include <getopt.h>
#include <signal.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#include "common.h"
#include "bubbles.h"


char *program_name = PROGRAM_NAME;
char *help_message = HELP_MESSAGE;


// in range 0.0 .. 1.0
float
frand ()
{
    return ((float) (rand() & 0377) / 256.0);
}


void
new (struct bubble *pb)
{
    int size;

    pb->y = LINES - 1;
    pb->x = (int) ((frand() * (float) (COLS - 1)) + 0.5);
    pb->speed = (int) (frand() * 3.0) + 1;
    size      = (int) (frand() * 4.0);

    switch (size) {
    case 1:
        pb->image = '.';
        break;

    case 2:
        pb->image = 'o';
        break;

    case 3:
        pb->image = 'O';
        break;

    default:
        pb->image = NOBUBBLE;
    }
}


void
pop (struct bubble *pb)
{
    switch (pb->image) {
    case '.':
        mvaddch(SURFACE, pb->x, '_');
        break;

    case 'o':
        mvaddch(SURFACE, pb->x, 'v');
        break;

    case 'O':
        mvaddch(SURFACE, pb->x, '*');
        mvaddch(SURFACE - 1, pb->x, '.');
        break;

    default:
        return;
    }

    pb->image = NOBUBBLE;
}


int
main (int argc, char *argv[])
{
    static struct bubble *bubbles;
    int c;
    int ch;
    long delay_count = DELAY_COUNT;
    long count = 0;
    struct timespec req = {
        .tv_sec = 0,
        .tv_nsec = 1000000
    };
    int quantity;
    int ix;
    struct option long_options[] = {
        {"delay"  , required_argument, NULL, 'd'},
        {"help"   , no_argument      , NULL, 'h'},
        {"version", no_argument      , NULL, 'v'},
        {NULL     , 0                , NULL,  0 }
    };

    while ((c = getopt_long(argc, argv, "d:hv", long_options, NULL)) != -1)
        switch (c) {
        case 'd':
            delay_count = atoi(optarg);
            break;

        case 'h':
            usage(EXIT_SUCCESS);
            break;

        case 'v':
            print_version();
            exit(EXIT_SUCCESS);

        default:
            usage(EXIT_FAILURE);
        }

    srand(time(NULL));
    initscr();
    noecho();
    curs_set(0);
    nonl();
    cbreak();
    nodelay(stdscr, true);
    signal(SIGTERM, abort_handle);
    signal(SIGINT,  abort_handle);

    quantity = COLS / 4;
    bubbles = (struct bubble *) calloc(quantity, sizeof(struct bubble));
    if (bubbles == NULL) {
        perror(argv[0]);
        exit(EXIT_FAILURE);
    }

    for (ix = 0; ix < quantity; ix++)
        bubbles[ix].image = NOBUBBLE;

    count = 0;
    for (;;) {
        ch = getch();
        if (ch == 'q')
            break;

        nanosleep(&req, NULL);
        if (--count > 0)
            continue;
        count = delay_count;

        erase();

        mvaddch(SURFACE, 0, '\\');
        for (ix = 1; ix < (COLS - 1); ix++)
            mvaddch(SURFACE, ix, '_');
        mvaddch(SURFACE, COLS - 1, '/');

        for (ix = 0; ix < quantity; ix++) {
            if (bubbles[ix].image == NOBUBBLE)
                new(&(bubbles[ix]));
            else {
                bubbles[ix].y -= bubbles[ix].speed;
                if (bubbles[ix].y < SURFACE)
                    pop(&(bubbles[ix]));
            }

            if (bubbles[ix].image != NOBUBBLE)
                mvaddch(bubbles[ix].y, bubbles[ix].x, bubbles[ix].image);

        }
        move(0, 0);
        refresh();
        nanosleep(&req, NULL);
    }
    endwin();
}


bubbles.h
/*
 * Copyright (C) 2014 Yu-Jie Lin
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#include "common.h"


#define PROGRAM_NAME "bubbles"
#define HELP_MESSAGE "\
Usage: " PROGRAM_NAME  " [OPTION]...\n\
Animated ASCII bubbles in terminal\n\
\n\
  -d, --dalay=DELAY         update interval in milliseconds\n\
                            (Default: " STR(DELAY_COUNT) ")\n\
  -h, --help                display this help and exit\n\
  -v, --version             display version and exit\n\
\n\
Report bugs to <" URL_ISSUES ">\n\
Home page: <" URL_HOMEPAGE ">\n\
For complete documentation, run: man " PROGRAM_NAME

#define SURFACE  ((LINES) / 5)
#define NOBUBBLE ' '

// macro helpers
#define STR(s)  XSTR(s)
#define XSTR(s) #s

// default option values
#define DELAY_COUNT 500


struct bubble {
    // 0 .. LINES
    int x;
    // 0 .. COLS
    int y;
    // '.', 'o', 'O': ' ' == no bubble
    char image;
    // 1 .. 3
    int speed;
};


clock.c
/*
 * ASCII analog clock in terminal
 * Copyright (C) 2014 Alexandre Dantas
 * Copyright (C) 2014 Yu-Jie Lin
 * Copyright (C) 1993 ZOIS Ltd.
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 1).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <curses.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>
#include <math.h>
#include <setjmp.h>
#include <signal.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <time.h>
#include <unistd.h>

#include "common.h"
#include "clock.h"


char *program_name = PROGRAM_NAME;
char *help_message = HELP_MESSAGE;

char *second_styles[] = {
    [SS_OFF ] = "OFF",
    [SS_BLOB] = "BLOB",
    [SS_LINE] = "LINE"
};
unsigned int n_second_styles = sizeof(second_styles) / sizeof(char *);

jmp_buf reset;


void
roman_clock_face (char *title, char *date)
{
    static int lines = 0;
    static int cols = 0;
    int dash;
    static char *number[12] = {
        "XII",
        "I",
        "II",
        "III",
        "IIII",
        "V",
        "VI",
        "VII",
        "VIII",
        "IX",
        "X",
        "XI"
    };
    static struct icartesian at[12];
    float x;
    float y;
    float vector;

    if (lines != LINES || cols != COLS) {
        lines = LINES;
        cols = COLS;

        for (dash = 0; dash < 12; dash++) {
            switch (dash) {
            case 0:
                x =  0.0;
                y = -1.0;
                break;
            case 3:
                x = 1.0;
                y = 0.0;
                break;
            case 6:
                x = 0.0;
                y = 1.0;
                break;
            case 9:
                x = -1.0;
                y =  0.0;
                break;
            default:
                vector = (M_PI / 6.0 * (float) dash) - (M_PI / 2.0);
                y = tan(vector);
                x = tan((M_PI / 2.0) - vector);

                if (vector < 0.0)
                    x *= -1.0;
                if (vector < M_PI && vector > (M_PI / 2.0))
                    y *= -1.0;
                if (vector > M_PI) {
                    x *= -1.0;
                    y *= -1.0;
                }

                if (y < -1.0)
                    y = -1.0;
                if (x < -1.0)
                    x = -1.0;
                if (y > 1.0)
                    y = 1.0;
                if (x > 1.0)
                    x = 1.0;
            }

            at[dash].ix = (int) ((x * HALF_X * ASPECT) + HALF_X);
            at[dash].iy = (int) ((y * HALF_Y)          + HALF_Y);

            if (at[dash].iy == 0 || at[dash].iy == (LINES - 1))
                at[dash].ix -= (int) strlen(number[dash]) / (int) 2;
            if (at[dash].ix >= (int) ((HALF_X * ASPECT) + HALF_X))
                at[dash].ix -= strlen(number[dash]) - 1;
        }
    }

    mvaddstr(
        (int) (LINES - 1) * (int) 3 / (int) 4,
        ((int) (COLS - 1) / (int) 2) - ((int) strlen(date) / (int) 2),
        date
    );

    mvaddstr(
        (int) (LINES - 1) / (int) 4,
        ((int) (COLS - 1) / (int) 2) - ((int) strlen(title) / (int) 2),
        title
    );

    for (dash = 0; dash < 12; dash++)
        mvaddstr(at[dash].iy, at[dash].ix, number[dash]);
}


void
clock_face (char *title, char *date)
{
    static int lines = 0;
    static int cols  = 0;
    int dash;
    static char dashes[] = "|//-\\\\|//-\\\\";
    float vector;
    static struct icartesian at[12];

    if (lines != LINES || cols != COLS) {
        lines = LINES;
        cols  = COLS;
        for (dash = 0; dash < 12; dash++) {
            vector = (M_PI / 6.0 * (float) dash) - (M_PI / 2.0);
            at[dash].iy = (int) ((sin(vector) * HALF_Y)          + HALF_Y);
            at[dash].ix = (int) ((cos(vector) * HALF_X * ASPECT) + HALF_X);
        }
    }

    if ((int) strlen(date) <= (int) COLS)
        mvaddstr(
            (int) ((LINES - 1) * 3) / (int) 4,
            ((int) (COLS - 1) / (int) 2) - ((int) strlen(date) / (int) 2),
            date
        );

    if ((int) strlen(title) <= (int) COLS)
        mvaddstr(
            (int) (LINES - 1) / (int) 4,
            ((int) (COLS - 1) / (int) 2) - ((int) strlen(title) / (int) 2),
            title
        );

    for (dash = 0; dash < 12; dash++)
        mvaddch(at[dash].iy, at[dash].ix, dashes[dash]);
}


/*
 * vector: radians
 * length: relative 0.0 .. 1.0
 */
void
drawline (double vector, double length, char c)
{
    struct cartesian end;
    struct cartesian start;

    // orientate 0 == straight up
    vector -= M_PI / 2.0;
    end.y = (sin(vector) * length * HALF_Y)          + HALF_Y;
    end.x = (cos(vector) * length * HALF_X * ASPECT) + HALF_X;
    start.y = HALF_Y;
    start.x = HALF_X;
    plot(&start, &end, c);
}


/*
 * vector: radians
 * length: relative 0.0 .. 1.0
 */
void
drawblob (double vector, double length, char c)
{
    vector -= M_PI / 2.0;
    mvaddch(
        (int) ((sin(vector) * length * HALF_Y)          + HALF_Y),
        (int) ((cos(vector) * length * HALF_X * ASPECT) + HALF_X),
        c
    );
}


void
plot (struct cartesian *start, struct cartesian *end, char c)
{
    putline(
        (int) start->x,
        (int) start->y,
        (int) end->x,
        (int) end->y,
        c
    );
}


/*
 * See Newman & Sproull "Principles of Interactive Computer
 * Graphics", McGraw-Hull, New York, 1979 pp 33-44.
 */
void
putline (int x0, int y0, int x1, int y1, char c)
{
    int dx;
    int dy;
    int a;
    int b;
    int two_a;
    int two_b;
    int xcrit;
    int eps;

    dx = 1;
    a = x1 - x0;
    if (a < 0) {
        dx = -1;
        a = -a;
    }

    dy = 1;
    b = y1 - y0;
    if (b < 0) {
        dy = -1;
        b = -b;
    }

    two_a = 2 * a;
    two_b = 2 * b;

    xcrit = -b + two_a;
    eps = 0;
    mvaddch(y0, x0, c);
    while  (x0 != x1 || y0 != y1) {
        if (eps <= xcrit) {
            x0 += dx;
            eps += two_b;
        }
        if (eps >= a || a <= b) {
            y0 += dy;
            eps -= two_a;
        }

        mvaddch(y0, x0, c);
    }
}


double
asrads (int degrees)
{
    return (((double) degrees / (double) 180.0) * (double) M_PI);
}


char *
get_date (time_t *pt)
{
    char *s;

    s = ctime(pt);
    s[10] = '\0';
    return s;
}


void
winch_handle ()
{
    struct winsize size;

    if (ioctl(2, TIOCGWINSZ, &size) >= 0) {
        LINES = size.ws_row;
        COLS  = size.ws_col;
    }
    endwin();
    longjmp(reset, 1);
}


void
my_clock (struct clock_mode *mode, char *title)
{
    struct  tm *t;
    time_t  tr;
    int     hour_hand;
    int     min_hand;
    char    *date;
    int     ch;
    // 1/100 seconds
    struct  timespec req = {
        .tv_sec = 0,
        .tv_nsec = 10000000
    };
    time_t  next_update = 0;

    initscr();
    noecho();
    curs_set(0);
    nonl();
    cbreak();
    nodelay(stdscr, true);
    signal(SIGINT,  abort_handle);
    signal(SIGTERM, abort_handle);

    for (;;) {
        ch = getch();
        if (ch == 'q')
            break;

        switch (ch) {
        case 's':
            mode->second = mode->second == SS_END ? SS_BEGIN : mode->second + 1;
            next_update = 0;
            break;

        case 'r':
            mode->roman = !mode->roman;
            next_update = 0;
            break;

        case 'f':
            mode->day_date = !mode->day_date;
            next_update = 0;
            break;

        case 'd':
            mode->digital = !mode->digital;
            next_update = 0;
            break;
        }

        nanosleep(&req, NULL);
        if (time(NULL) >= next_update) {
            time(&tr);
            t = localtime(&tr);

            if (mode->second != SS_OFF) {
                next_update = time(NULL) + 1;
            } else {
                // sync up to min
                time(&tr);
                t = localtime(&tr);
                next_update = time(NULL) + (60 - t->tm_sec);
            }
        } else
            continue;

        erase();

        if (mode->day_date)
            date = get_date(&tr);
        else if (mode->digital)
            date = ctime(&tr);
        else
            date = "";

        if (mode->roman)
            roman_clock_face(title, date);
        else
            clock_face(title, date);

        hour_hand = (
            (t->tm_hour % 12) * (360 / 12)) +
            (t->tm_min * 360 / 60 / 12) +
            (t->tm_sec * 360 / 60 / 60 / 12
        );
        drawline(asrads(hour_hand), 0.6, '#');

        min_hand = (t->tm_min * 360 / 60) + (t->tm_sec * 360 / 60 / 60);
        drawline(asrads(min_hand), 0.9, '*');

        switch (mode->second) {
        case SS_OFF:
            break;

        case SS_LINE:
            drawline(asrads(t->tm_sec * 360 / 60), 0.7, '@');
            break;

        default:
            drawblob(asrads(t->tm_sec * 360 / 60), 0.7, '@');
            break;
        }

        move(0, 0);
        refresh();
    }

    endwin();
}


int
main (int argc, char *argv[])
{
    int  c;
    long i;
    char *title;
    char *ns;
    char *end;

    struct clock_mode mode = {
        .second   = SS_DEFAULT,
        .roman    = false,
        .day_date = false,
        .digital  = false
    };
    struct option long_options[] = {
        {"second" , optional_argument, NULL, 's'},
        {"roman"  , no_argument      , NULL, 'r'},
        {"date"   , no_argument      , NULL, 'd'},
        {"help"   , no_argument      , NULL, 'h'},
        {"version", no_argument      , NULL, 'v'},
        {NULL     , 0                , NULL,  0 }
    };

    setlocale(LC_ALL, "");
    while ((c = getopt_long(argc, argv, "s::rfdhv", long_options, NULL)) != -1)
        switch (c) {
        case 's':
            if (optarg == NULL) {
                mode.second = SS_DEFAULT_ON;
                break;
            }

            i = strtol(optarg, &end, 10);
            if (end == optarg || *end != '\0' || errno == ERANGE) {
                for (i = 0; (unsigned) i < n_second_styles; i++)
                    if (strcmp(optarg, second_styles[i]) == 0)
                        break;
                if (i <= SS_END) {
                    mode.second = i;
                    break;
                }
            } else {
                mode.second = atoi(optarg);
            }
            break;

        case 'r':
            mode.roman = true;
            break;

        case 'f':
            mode.day_date = true;
            break;

        case 'd':
            mode.digital = true;
            break;

        case 'h':
            usage(EXIT_SUCCESS);
            break;

        case 'v':
            print_version();
            exit(EXIT_SUCCESS);

        default:
            usage(EXIT_FAILURE);
        }

    title = DEFAULT_TITLE;
    if (optind < argc)
        title = strdup(argv[optind++]);
    for (; optind < argc; optind++) {
        ns = malloc(strlen(title) + strlen(argv[optind]) + 2);
        if (ns == NULL) {
            perror(argv[0]);
            exit(EXIT_FAILURE);
        }
        strcpy(ns, title);
        strcat(ns, " ");
        strcat(ns, argv[optind]);
        free(title);
        title = ns;
    }

    setjmp(reset);
    signal(SIGWINCH, winch_handle);
    my_clock(&mode, title);
    exit(EXIT_SUCCESS);
}
